RedHatInsights/rhsm-subscriptions
# ENT-3667: lookup orgId when performing auto-opt-in

I also added a few corresponding methods to the opt-in JMX bean.

The lookup only happens if the account doesn't already have an opt-in record.

Testing
-------

1. Clear the account_config and org_config tables:

```sql
truncate table account_config;
truncate table org_config;
```

2. Run the application with modified PromQL and connected to the QA user
   service - replace `$RHSM_KEYSTORE` and `$RHSM_KEYSTORE_PASSWORD`:

Run a proxy to prometheus per (internal) https://docs.engineering.redhat.com/display/ENT/Connecting+to+Observatorium+to+Fetch+Prometheus+Metrics

```
DEV_MODE=true \
  OPENSHIFT_MAX_ATTEMPTS=1 \
  RHSM_SUBSCRIPTIONS_METERING_PROMETHEUS_METRIC_OPENSHIFT_ENABLED_ACCOUNT_PROM_Q_L="group(min_over_time(subscription_labels{ebs_account != '', billing_model='standard'}[1h])) by (ebs_account)" \
  PROM_URL=http://localhost:8082/api/v1 \
  USER_HOST=user.qa.api.redhat.com \
  RHSM_KEYSTORE=$RHSM_KEYSTORE \
  RHSM_KEYSTORE_PASSWORD=$RHSM_KEYSTORE_PASSWORD \
  ./gradlew bootRun
```

3. Invoke the JMX API to get the service to tally accounts.

```
curl 'http://localhost:8080/actuator/hawtio/jolokia/' -H 'Content-Type: application/json' -d '{"type":"exec","mbean":"org.candlepin.subscriptions.metering.jmx:name=openshiftJmxBean,type=OpenShiftJmxBean","operation":"performOpenshiftMetering()","arguments":[]}'
```

It is normal to see some warnings about account not being found due to data differences between prometheus and the user API.

4. View resulting records in `account_config` and `org_config` tables.

```sql
select * from org_config;
select * from account_config;
````

5. Try some of the new methods on the OptInJmxBean if desired (http://localhost:8080/actuator/hawtio) Hint: you can use the log messages to get some sample org IDs and account numbers.
